### PR TITLE
refactor AttributeSelectorPopup code, add some tests

### PR DESCRIPTION
**Pull request by zkrejcov**
The original AttributeSelectorPopup code was almost 1:1 copied to GuidedDecisionTableViewImpl. This removes the duplication - the AttributeSelectorPopup is now an abstract parent of 2 separate implementations. Shared code (most logic) stayed there.

This also brings the basic metadata input validation that was present in GRE to GDT.

There are now also a few small tests checking that validation and that duplicates are not allowed (part of moving selenium tests to community).

### CODE REVIEW

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableAttributeSelectorPopup.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        this.existingAttributeNames = existingAttributeNames;
+        this.presenter = presenter;
+
+        initialize();
+    }
+
+    @Override
+    protected String[] getAttributes() {
+        String[] attributes = RuleAttributeWidget.getAttributesList();
+        attributes[attributes.length] = GuidedDecisionTable52.NEGATE_RULE_ATTR;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableAttributeSelectorPopup.java]: Isn't this wrong?

It should either be ```attributes[attributes.length-1]``` (and overwrite the last entry in the array, which is probably equally wrong) or it should extend the array. The old code used a ```List``` so we could append an additional attribute to the end (Decision Tables support a "negate rule" attributed, Guided Rules do not).

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopup.java [hunk 2]
```diff
@@ -16,81 +16,63 @@
...
+        for ( String item : getAttributes() ) {
+            list.addItem( item );
+        }
 
         // Remove any attributes already added
-        for ( RuleAttribute at : model.attributes ) {
+        for ( String at : getDuplicates() ) {
             for ( int iItem = 0; iItem < list.getItemCount(); iItem++ ) {
-                if ( list.getItemText( iItem ).equals( at.getAttributeName() ) ) {
+                if ( list.getItemText( iItem ).equals( at ) ) {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopup.java]: Why check if the attribute names are equal before removing, when the method you called is ```getDuplicates()``` which (in my mind) suggests the collection already contains the attributes to remove?

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopup.java [hunk 3]
```diff
@@ -16,81 +16,63 @@
...
         addAttribute( GuidedRuleEditorResources.CONSTANTS.Metadata3(),
                       horiz );
 
     }
 
-    private void setListBox( final RuleModel model,
-                             final Command refresh ) {
-
-        final ListBox list = RuleAttributeWidget.getAttributeList();
+    private void setAtributesPanel() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopup.java]: Typo, two "T"'s in "at*t*ributes"

**Comment by manstis** [comment]
@zkrejcov Just a couple of minor comments - well, one less minor than the other ;)

If you could fix these (and possibly a test for the issue that is likely to throw a runtime exception?) it'd be "lovely" ;)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableAttributeSelectorPopup.java (hunk 1 above)
Comment (triceo) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableAttributeSelectorPopup.java]: Nice catch! This should have thrown `ArrayIndexOutOfBoundsException`, should it not? And the fact that we do not know about it would indicate that either the code is not used (doubt that), or it is not covered by tests.

Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableAttributeSelectorPopup.java]: Yes it would have thrown an ```ArrayIndexOutOfBoundsException``` if the PR was merged "as is".. it didn't fail in the old code as we used a ```List``` not an array. It is probably not covered by your tests.

Comment (zkrejcov) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableAttributeSelectorPopup.java]: Thanks for catching that, I'll push a fixed commit soon.

**Comment by zkrejcov** [comment]
jenkins please retest this

**Comment by zkrejcov** [comment]
"[ERROR] Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x0000000763900000, 477102080, 0) failed; error='Cannot allocate memory' (errno=12)" when building Drools Workbench - WebApp

jenkins please retest this

**Comment by zkrejcov** [comment]
@manstis could you please check if it is okay now? That pesky issue should be fixed and a test was added.
I'm partly expecting to learn that GWT doesn't support this or that either (e.g. Arrays.copyOf(..)) ever since that small String.format(..) surprise. ;)

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/test/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopupTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    private static class MockAttributeSelectorPopup extends AttributeSelectorPopup {
+
+        @Override
+        protected String[] getAttributes() {
+            return attributes;
+        }
+
+        @Override
+        protected String[] getDuplicates() {
+            System.out.println("getting duplicates");
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/test/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopupTest.java]: Sorry, @zkrejcov I missed this the first time round.. it'd be best to remove this type of logging (which I assume you added to help debug your tests, or the like)

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/GuidedRuleAttributeSelectorPopup.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import org.drools.workbench.screens.guided.rule.client.resources.GuidedRuleEditorResources;
+import org.gwtbootstrap3.client.ui.Button;
+import org.uberfire.ext.widgets.common.client.common.InfoPopup;
+
+public class GuidedRuleAttributeSelectorPopup extends AttributeSelectorPopup {
+
+    private final RuleModel model;
+    private final Command refresh;
+
+    public GuidedRuleAttributeSelectorPopup( final RuleModel model,
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/GuidedRuleAttributeSelectorPopup.java]: I know it's a copy and paste; but could you make the parameter types consistent; either all ```final``` or none ```final```?

**Comment by zkrejcov** [comment]
Hmm, oops? :) That 'println()' should not have stayed there. Okay, hopefully now good.

I added 'final' to the other parameters - more consistent with the current code style of these bits.

**Comment by zkrejcov** [comment]
I _think_ the fail is due to https://github.com/droolsjbpm/kie-wb-common/pull/388 and will be gone with https://github.com/droolsjbpm/drools-wb/pull/204 ...

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
@zkrejcov Sorry.. looks like this has merge conflicts now some of the other PRs have been merged :-(

**Comment by zkrejcov** [comment]
jenkins please retest this

**Comment by psiroky** [comment]
The test failures are not related to this PR, they are failing on master as well

**Comment by zkrejcov** [comment]
Test fails - org.drools.workbench.screens.guided.dtable.client.widget.table.model.linkmanager.impl.DefaultGuidedDecisionTableLinkManagerTest @manstis I belive you added that recently..? It complains that "Mockito can only mock visible & non-final classes" and the line it complains about has `spy( wrapped )` where `final DefaultGuidedDecisionTableLinkManager wrapped` ...

**Comment by manstis** [comment]
@zkrejcov Probably need this https://github.com/droolsjbpm/drools-wb/pull/208... being reviewed and merged as I type....

**Comment by zkrejcov** [comment]
@manstis great, thanks for the info.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
+1 to merge.
